File: Trabalhos/2 - PontoFlutuante/Pontoflutuante_cpp/MATA49_PontoFlutuante_Main.cpp
/*
 * MATA49_PontoFlutuante_Main.cpp
 *
 *  Created on: 8 de dez de 2017
 *      Author: raffaello
 */

#include <iostream>
#include "funcao/assembly/MyAsmFunctions.h"
#include "funcao/base/FuncaoUnaria.h"
#include "funcao/base/TimeWorks.h"

using namespace std;

extern FuncaoUnaria<double, double> *_dummy;
extern FuncaoUnaria<double, double> *_abs;
extern FuncaoUnaria<double, double> *_exp;
extern FuncaoUnaria<double, double> *_ln;
extern FuncaoUnaria<double, double> *_sqrt;
extern FuncaoUnaria<double, double> *_cos;
extern FuncaoUnaria<double, double> *_sin;
extern FuncaoUnaria<double, double> *_tan;
extern FuncaoUnaria<double, double> *_acos;
extern FuncaoUnaria<double, double> *_asin;
extern FuncaoUnaria<double, double> *_atan;

int main(int argc, char* argv[]) {
	FuncaoUnaria<double, double> *funcoes[] = {
			_dummy,
			_abs,
			_exp,
			_ln,
			_sqrt,
			_cos,
			_sin,
			_tan,
			_acos,
			_asin,
			_atan
	};

	double x = 30;
	if(argc > 1) {
		cout << "parametro de entrada: " << argv[1] << endl;
		x = atof(argv[1]);
	}
	cout << "inicio" << endl;
	for (FuncaoUnaria<double, double> *fu : funcoes) {
		if(fu != NULL) {
			double outCpp = fu->operarCpp(x);
			double outAsm = fu->operarAsm(x);
			TimeWorks *tw = (dynamic_cast<TimeWorks*>(fu));
			cout << fu->getName() << "(" << x << ")" << endl;
			cout << "\tResultado C++: " << outCpp << "; Tempo C++: " << tw->getExecTimeCpp() << endl;
			cout << "\tResultado Asm: " << outAsm << ";Tempo Asm: " << tw->getExecTimeAsm() << endl;
			cout << "\tGanho: " << tw->getExecTimeCpp().count() << " / " << tw->getExecTimeAsm().count() << " = " << (tw->getExecTimeCpp().count() / (tw->getExecTimeAsm().count() == 0 ? 0.00001 : tw->getExecTimeAsm().count())) << endl;
			cout << endl;
		}
	}
	cout << "fim" << endl;
	system("pause");
}

File: Trabalhos/2 - PontoFlutuante/PontoFlutuante_Java/src/br/ufba/mata49/trabalho/pontoflutuante/MATA49_PontoFlutuante_Main.java
package br.ufba.mata49.trabalho.pontoflutuante;

import java.util.Arrays;

import br.ufba.mata49.trabalho.pontoflutuante.funcao.Abs;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.Acos;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.Asin;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.Atan;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.Cos;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.Dummy;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.Exp;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.Ln;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.Sin;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.Sqrt;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.Tan;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.base.FuncaoUnaria;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.base.TimeWorks;

public class MATA49_PontoFlutuante_Main {

	private FuncaoUnaria funcoes[] = { 
				new Dummy(),
				new Abs(),
				new Exp(),
				new Ln(),
				new Sqrt(),
				new Cos(),
				new Sin(),
				new Tan(),
				new Acos(),
				new Asin(),
				new Atan()
			};
	
	public MATA49_PontoFlutuante_Main(String[] args) {
		double x = 30;
		if(args.length > 0) {
			System.out.println("parametro de entrada: " + args[0]);
			x = Double.parseDouble(args[0]);
		}
		for (FuncaoUnaria fu : Arrays.asList(funcoes)) {
			System.out.println(fu.getName() + "(" + x + ")");
			System.out.println("\tResultado Java: " + fu.operar(x) + "; Tempo: " + ((TimeWorks)fu).getExecTime() + " nanoseconds");
			System.out.println("");
		}
	}
	
	public static void main(String[] args) {
		new MATA49_PontoFlutuante_Main(args);
	}
}
